design-document/design-document.md

```diff
@@ -0,0 +1,131 @@
+## Frontend
+
+Let's use React (`create-react-app` for the simplicity of its setup), React Router, and Tailwind on the frontend. Core considerations:
+
+ - **Authentication:** All protected routes will be wrapped in a simple component `<Protected />` which checks whether the user is authenticated. When they're not, the user is taken to the `/sign-in` page, and the original URL is stored in the `state` property on the `location` object of our router. That way, we can redirect the user back to where they intended to go once they've signed in.
+ - **URL as the source of truth:** We will store user's current working directory inside the file tree as URL params, the sorting and filtering criteria as query params (`/browse/my-directory/my-child-directory?sort=name-up&search=holida`).
+ - **Error handling**: Let's pay attention to loading states, errors (at the very least 400, 401, 403, 404, 500) and edge cases such as invalid filtering options, etc.
+ - **Security:** Input and output shall be sanitised (URL params, query params, and the search input field).
+
+## Backend
+
+The server will be a plain `express` app. For the sake of simplicity, let's skip a database and store everything in memory instead.
+
+Since searching and filtering is required on the client only, a handful of API endpoints will suffice:
+
+ - `GET /contents?path={path}` to obtain 1-level deep contents of a directory, returning `{ contents: Item }` where `interface Item { name: string, sizeKb: number | undefined, type: "file" | "dir", items: undefined | Item[] }`
+ - `POST /sign-in` to sign the user in
+ - `POST /sign-out` to sign the user out (`POST` to avoid [prefetching it](https://twitter.com/nick_craver/status/296281730984316928?lang=fa) 🤦) 
+
+## Security
+
+_Note: Whenever generating random, secure tokens, let's use [crypto.randomBytes()](https://nodejs.org/api/crypto.html#cryptorandombytessize-callback)._
+
+Security is, naturally, a huge consideration.
+
+### TLS
+
+To start with, let's use a self-signed certificate to enable `TLS` even on localhost. Unfortunately, this does imply you will need to import the certificate as trusted or "Accept the risk" in your browser. Cheers for your patience in advance 🙏
+
+### Session Management
+
+To avoid depending on external libs, we'll build a simple **session management** ourselves. On signing in, we generate a random `sessionId` for the user. This is used as the `key` in the `sessions` key-value store and also set as a cookie sent back to the client. (See more on security considerations in the [XSS](#cross-site-scripting-xss) and [CSRF](#cross-site-request-forgery-csrf) sections below.)
+
+### Directory Traversing
+
+Since we'll be browsing real directories, we also need to make sure the user is not allowed outside their root directory. After sanitising the URL input, [path.normalize()](https://nodejs.org/api/path.html#pathnormalizepath) followed by removing any remaining `./` and `../` at the beginning of the path should suffice.
```

Comment:
Normalize should remove any `.` and `..` right? So the last check would be checking if the root is what you expect?

Reply:
Yeah, only `.` and `..` at the beginning are left, e.g. `path.normalize("../parent/../parent-2/../parent-3/child")` returning `../parent-3/child`. My idea was to strip these should they exist and `path.join` with the root, but a more reasonable approach might be to return a `400` — it's not just access denied but a malformed request.

Reply:
Ah, I just tried `path.normalize` with NodeJS and see now. What do you think about something like:

``
path := basePath + normalize(requestedPath)
if !startsWith(path, basePath) {
   return an error
}
```

Reply:
I like that! My code (not yet in a PR) is:

```
const normalizedPath = path.normalize(rawPath);
if (normalizedPath.startsWith(".")) {
  return res.status(400).json({ message: "Invalid path." });
}
const parent = path.join(basePath, normalizedPath);
```

... so pretty similar, but yours is a step shorter and there's one less variable 👍